### PR TITLE
proof of concept code for RHAIENG-5450

### PR DESCRIPTION
**Pull request by kami619**
## Description

Implemented ServiceAccount token auth validation for `react_agent` PoC under `RHAIENG-4859` with per-agent audience and explicit caller allowlist authorization.

### What changed

- Added shared auth package: `components/auth`
  - `SATokenAuthMiddleware` with:
    - `TokenReview` authn
    - audience check (`AUTH_AUDIENCE`)
    - ServiceAccount allowlist authz (`AUTH_ALLOWED_SERVICEACCOUNTS`)
    - status behavior: `401` (missing/invalid/wrong audience), `403` (not allowlisted), `503` (TokenReview infra failure)
- Wired middleware into `react_agent`
  - `main.py`, `pyproject.toml`, `Dockerfile`, `Makefile`, `.env.example`, `agent.yaml`, `values.yaml`
- Added Helm/chart support for auth-based deployment
  - `serviceaccount` template
  - `auth-delegator` ClusterRoleBinding template
  - deployment env wiring for auth vars
  - shared auth defaults in chart values (`auth.enabled: false`)
- Added tests for auth-based validation
  - Unit tests: `agents/langgraph/react_agent/tests/test_auth_middleware.py`
  - Integration auth matrix: `agents/langgraph/react_agent/tests/integration/test_auth.py`
  - Integration helpers/fixtures updates in `tests/integration/utils.py` and `agents/langgraph/react_agent/tests/integration/conftest.py`

## Jira Ticket

RHAIENG-4859
RHAIENG-5450

## Testing

- [x] `make test` passes (run from the affected agent directory)
- [x] Manual testing performed (describe steps below)
- [ ] No testing required (documentation/config change only)

### Additional verification steps performed

- `cd agents/langgraph/react_agent && make test`  
  - Result: `17 passed`
- `cd agents/langgraph/react_agent && uv run --extra dev python -m pytest tests/test_auth_middleware.py -q`  
  - Result: `7 passed`
- `cd agents/langgraph/react_agent && BASE_URL=http://vllm-svc.llama-serving.svc.cluster.local:8000/v1 MODEL_ID=qwen2-5-7b-instruct API_KEY=not-needed PYTHONPATH="$(git rev-parse --show-toplevel)/tests" uv run --extra dev python -m pytest tests/integration/test_auth.py -v --tb=long`  
  - Result: `1 passed` (`test_auth_enforcement_matrix`)
- Manual endpoint verification after deploy:
  - no token -> `401`
  - allowlisted caller token -> `200`

## Checklist

- [x] I have read [CONTRIBUTING.md](CONTRIBUTING.md)
- [x] No `.env` or secret files are included in this PR
- [x] All changes are within scope of the linked Jira ticket (if not, explain in Description)

## Review Guidance

- Focus review on:
  - `components/auth/src/agent_auth/middleware.py` (core authn/authz behavior)
  - `charts/agent/templates/deployment.yaml` + new auth RBAC templates
  - `agents/langgraph/react_agent/tests/integration/test_auth.py` (expected `401/403/200` matrix)
- Note: shared chart default remains auth-off; PoC enablement is agent-scoped.

## Related PRs

None.

### CODE REVIEW

File: agents/langgraph/react_agent/values.yaml [hunk 1]
```diff
@@ -3,3 +3,9 @@ nameOverride: "langgraph-react-agent"
 env:
   MODEL_ID: ""
   BASE_URL: ""
+
+auth:
+  enabled: true
+  audience: "langgraph-react-agent"
+  allowedServiceAccounts: []
+  createAuthDelegatorBinding: true
```
[the remark this answers is not in the document]
Comment (kami619) [reply agents/langgraph/react_agent/values.yaml]: you missed that the shared chart (charts/agent/values.yaml:69) defaults createAuthDelegatorBinding: false. The agent-specific values.yaml sets it true because this agent needs it — that's the whole point of the PoC. Anyone deploying this specific agent needs the CRB. The shared chart is safe for other agents.

File: charts/agent/templates/deployment.yaml [hunk 2]
```diff
@@ -37,6 +40,16 @@ spec:
...
             {{- end }}
+            {{- if .Values.auth.enabled }}
+            - name: AUTH_ENABLED
+              value: "true"
+            - name: AUTH_AUDIENCE
+              value: {{ .Values.auth.audience | quote }}
+            {{- if .Values.auth.allowedServiceAccounts }}
+            - name: AUTH_ALLOWED_SERVICEACCOUNTS
+              value: {{ join "," .Values.auth.allowedServiceAccounts | quote }}
+            {{- end }}
```
[the remark this answers is not in the document]
Comment (kami619) [reply charts/agent/templates/deployment.yaml:51]: The agent-level values.yaml sets audience: "langgraph-react-agent" and allowedServiceAccounts is intentionally configurable (populated at deploy time via .helm-secrets.yaml). Forcing a non-empty list at render-time would break the "deploy with auth enabled, configure callers later" workflow. The AUTH_AUDIENCE is already set; an empty allowlist just means deny-all, which is a valid (if strict) default.

File: components/auth/src/agent_auth/middleware.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+            config.load_incluster_config()
+            self._auth_api = client.AuthenticationV1Api()
+        return self._auth_api
+
+    def _validate_token(self, token: str) -> Any | None:
+        review = self._get_auth_api().create_token_review(
+            client.V1TokenReview(
+                spec=client.V1TokenReviewSpec(token=token, audiences=[self.audience])
+            )
+        )
```
[the remark this answers is not in the document]
Comment (kami619) [reply components/auth/src/agent_auth/middleware.py:101]: _validate_token is a synchronous K8s API call made from an async ASGI __call__. Under load, this blocks the event loop for every request. The suggestion to
  use anyio.to_thread.run_sync and add _request_timeout is technically correct. However — for a PoC with low concurrency this is unlikely to hit this here., I will do it as a follow up once this becomes the official path to securing the agents.

File: agents/langgraph/react_agent/tests/integration/conftest.py [hunk 4]
```diff
@@ -1,2 +1,144 @@
...
+        run_make("build-openshift", cwd=agent_dir, timeout=1200)
+
+        logger.info("Deploying to cluster...")
+        run_make("deploy", cwd=agent_dir, timeout=300)
+        deployed = True
+
+        route_url = get_route(agent_name, namespace=namespace)
+        logger.info("Agent deployed at %s", route_url)
+        yield route_url
+    except (MakeTargetError, RouteNotFoundError) as exc:
```
[the remark this answers is not in the document]
Comment (kami619) [reply agents/langgraph/react_agent/tests/integration/conftest.py]: I will revisit this when the tests are actually flaky. closing for now.

File: components/auth/README.md [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+### 2) Deploy agent with auth enabled
+
+From `agents/langgraph/react_agent/.env`:
+
+```dotenv
+AUTH_ENABLED=true
+AUTH_AUDIENCE=langgraph-react-agent
+AUTH_ALLOWED_SERVICEACCOUNT=ci-testing:langgraph-react-agent-caller
+```
```
[the remark this answers is not in the document]
Comment (kami619) [reply components/auth/README.md:60]: fixed in https://github.com/red-hat-data-services/agentic-starter-kits/pull/158/commits/685fe1cb69d934344cb0217dfd8d9832e6577a4c

File: charts/agent/templates/deployment.yaml [hunk 6]
```diff
@@ -18,6 +18,9 @@ spec:
     spec:
       securityContext:
         runAsNonRoot: true
+      {{- if .Values.auth.enabled }}
+      serviceAccountName: {{ .Values.serviceAccount.name | default (include "agent.fullname" .) }}
+      {{- end }}
```
[the remark this answers is not in the document]
Comment (kami619) [reply charts/agent/templates/deployment.yaml:23]: The shared chart defaults have serviceAccount.create: true, so the fallback to agent.fullname will always point to the SA created by the new serviceaccount.yaml template. The failure mode you describe (create=false + name unset) isn't reachable with the current values. If someone deliberately sets create: false without setting name, that's operator error. A fail guard is nice-to-have Helm hygiene but not a bug in this PR.

**Review by andrewdonheiser — CHANGES_REQUESTED** [review]
## PR Review — ServiceAccount Token Auth POC

### Strengths

- **Fail-closed design.** The middleware defaults to rejecting requests. Missing/invalid/wrong-audience tokens and non-allowlisted callers are all explicitly blocked before the app is reached.
- **No token logging.** Token values never appear in log output. Error messages are generic and don't leak internal state.
- **Clean separation.** Extract/validate/authorize steps are clearly separated in the middleware. Status codes are correct and well-documented (401/403/503).
- **Solid test matrix.** Both unit tests (7 tests) and integration tests (4-case auth matrix + health bypass) are thorough for a POC.
- **CRB namespace-scoping.** `sha256sum .Release.Namespace | trunc 8` prevents ClusterRoleBinding name collisions across namespaces.
- **Shared chart guards.** `auth.enabled: false` and `createAuthDelegatorBinding: false` defaults mean existing agents are unaffected by the auth templates.
- **`system:auth-delegator` is minimum privilege.** Grants only TokenReview access, nothing else.

---

### Critical (1)

#### 1. Env var name mismatch: singular vs plural

`.env.example` and `agent.yaml` use `AUTH_ALLOWED_SERVICEACCOUNT` (singular), but the middleware reads `AUTH_ALLOWED_SERVICEACCOUNTS` (plural). The Helm pipeline bridges this correctly for cluster deploys, but **local dev with `AUTH_ENABLED=true` will silently 403 all requests** because the middleware reads the wrong env var name and gets an empty allowlist.

**Fix:** Rename to plural in `.env.example` and `agent.yaml` to match the middleware.

---

### Important (4)

#### 2. POC should not alter the non-auth path

This is a POC scoped to one agent, but the PR weaves auth into the base path instead of layering on top. Several changes affect `react_agent` (and in the case of the shared chart, **all** agents) even when auth is never enabled:

- **Hard dependency:** `agent-auth` is in `dependencies` (not `optional-dependencies`), so every `make env` installs `kubernetes>=29.0` (~40MB) even for local dev where auth is always off.
- **Unconditional middleware:** `app.add_middleware(SATokenAuthMiddleware)` is always registered. Every request traverses the middleware and the `kubernetes` module is imported at startup, even when `AUTH_ENABLED=false`.
- **Shared chart default:** `serviceAccount.create: true` means every agent deployed with the shared chart now gets a ServiceAccount created, even agents that never use auth.
- **Existing test rewrite:** The self-contained `deployed_agent` fixture in `test_deployment.py` was deleted and replaced with one that depends on `auth_callers` (creates 2 ServiceAccounts on the cluster). The health-check smoke test now requires auth infrastructure to run. If SA creation fails, the health test is blocked for an unrelated reason.

**Fix:** Layer the POC on top of the existing behavior:
- Make `agent-auth` an optional dependency (`[auth]` extra)
- Conditionally import and register the middleware only when `AUTH_ENABLED=true`
- Default `serviceAccount.create` to `false` in the shared chart
- Keep the existing `deployed_agent` fixture and `test_deployment.py` intact; have `test_auth.py` use its own auth-aware fixture

#### 3. Hardcoded default audience in shared middleware

`os.getenv("AUTH_AUDIENCE", "langgraph-react-agent")` — a shared component defaults to a specific agent's name. When a second agent adopts this without setting `AUTH_AUDIENCE`, it silently inherits react_agent's audience, undermining per-agent isolation.

**Fix:** Default to `""` and error/warn at startup when `AUTH_ENABLED=true` but `AUTH_AUDIENCE` is unset.

#### 4. Empty allowlist + auth enabled = silent denial of all traffic

`values.yaml` sets `auth.enabled: true` with `allowedServiceAccounts: []`. The service boots, passes health checks, but 403s every legitimate request. Fail-closed is correct, but this is a silent operational trap.

**Fix:** Add a startup log warning when `self.enabled and not self.allowed_serviceaccounts`, or add a Helm `fail` guard requiring at least one SA when auth is enabled.

#### 5. Dockerfile copies to absolute path outside WORKDIR

`COPY components/auth/ /components/auth/` places the package at filesystem root, outside the standard WORKDIR (`/opt/app-root/src`). Other COPY commands use relative paths. The directory also remains as dead weight after pip install.

**Fix:** `COPY components/auth/ ./components/auth/` (relative to WORKDIR), then `rm -rf ./components/auth` after install.

---

### Suggestions (4)

#### 6. `_run_oc_command` doesn't redact outputs

Unlike `run_make` which wraps errors in `_redact()`, the new `_run_oc_command` helper could leak cluster info or tokens into CI logs when it raises `RuntimeError`.

**Fix:** Wrap stderr/stdout in `_redact()` in the error message.

#### 7. `components/` directory pattern undocumented

This is the first shared cross-agent code in the repo. AGENTS.md says "keep agents self-contained — never import from another agent's `src/`". A `components/` directory is a reasonable evolution, but the pattern should be documented before more components are added.

**Fix:** Add `components/` to the repo structure docs in AGENTS.md and define what qualifies as a component vs. agent-level code.

#### 8. Defense-in-depth: verify audience in TokenReview response

The middleware trusts the K8s API server to enforce audience. A cheap additional check: `self.audience in review.status.audiences` after a successful review. If the API server ever has a misconfigured webhook, this catches it.

#### 9. Unit tests for `_caller_identity` edge cases

The static method is public and easy to test in isolation. Edge cases like trailing colons, extra colons, or non-SA usernames could be covered directly.

---

### Verdict: Request Changes

Two items should be fixed before merge:
1. **Env var naming mismatch** (Critical) — will cause silent failures in local dev
2. **POC should not alter the non-auth path** (Important) — the POC should layer on top of existing behavior without modifying the build, runtime, chart defaults, or test infrastructure for the non-auth case

The remaining Important items (hardcoded audience, empty allowlist trap, Dockerfile path) are acceptable for a POC if tracked for resolution before other agents adopt this pattern.

**Comment by kami619** [comment]
Thanks @andrewdonheiser appreciate your feedback, I will address them and get back with any questions/clarifications/fixes that I would have.

**Comment by kami619** [comment]
Addressed review `#pullrequestreview-4450916033`.

Implemented fixes:
- Fixed env var mismatch by standardizing on `AUTH_ALLOWED_SERVICEACCOUNTS` in local/runtime config, with backward-compatible Makefile fallback to singular.
- Made auth POC opt-in instead of baseline path changes:
  - moved `agent-auth` to optional dependency (`[project.optional-dependencies].auth`)
  - conditionally import/register middleware only when `AUTH_ENABLED=true`
  - set shared chart default `serviceAccount.create: false`
  - set `react_agent` default `auth.enabled: false`
  - restored deployment smoke test independence and moved auth-aware setup into `test_auth.py`
- Removed hardcoded audience default in shared middleware (`AUTH_AUDIENCE` now required when auth is enabled).
- Added startup warning when auth is enabled but allowlist is empty.
- Updated Dockerfile auth component copy/install path to be relative to `WORKDIR` and cleaned staged component files post-install.
- Redacted `_run_oc_command` error stdout/stderr in integration utilities.
- Documented `components/` conventions in `AGENTS.md`.
- Added defense-in-depth audience verification against TokenReview response audiences.
- Added `_caller_identity` edge-case tests and stricter parsing for malformed identities.

Verification run:
- `uv run --extra dev --extra auth python -m pytest tests/test_auth_middleware.py` -> 13 passed
- Lint diagnostics on touched files -> no issues

**Review by andrewdonheiser — APPROVED** [review]
👍
